pkg/hubble/cell/hubble.go

```diff
@@ -596,3 +603,17 @@ func getPort(addr string) (int, error) {
 	}
 	return portNum, nil
 }
+
+// Start implements cell.HookInterface.
+func (h *hubbleIntegration) Start(cell.HookContext) error {
+	go h.Launch(h.ctx)
```

Comment:
IMO keeping context as fields isn't a good approach. I'd refactor this and keep the lifecycle aspects within the cell instead of letting `hubbleIntegration` implement `HookInterface`.

->

```
	ctx, cancel := context.WithCancel(context.Background())

	p.Lifecycle.Append(cell.Hook{
		OnStart: func(_ cell.HookContext) error {
			h.Launch(ctx)
			return nil
		},
		OnStop: func(_ cell.HookContext) error {
			cancel()
			return nil
		},
	})

```

alternatively / as a follow up you could also take a look into the Hive job framework where you could use `OneShot` jobs where you also provide more insights into what jobs are running (instead of just spawning raw Go routines)

Reply:
I would highly recommend to use the jobs whenever possible. thanks!

Reply:
Nice, thanks a lot!

Just mentioning that it's not necessary to handle the health state explicitly in this case, as the running and error case are already covered the oneshot job internally. (returning an [error results in a degradation](https://github.com/cilium/hive/blob/main/job/oneshot.go#L148-L152) - otherwise [it's still in Running state](https://github.com/cilium/hive/blob/main/job/oneshot.go#L136))

The health checker could be used to pass on to the actual job logic and report change of state while the one shot job is still running. Probably out of scope but still good to know.

E.g. changing the job description to the following should be enough

```
	jobGroup.Add(job.OneShot("hubble", func(ctx context.Context, _ cell.Health) error {
		h.launch(ctx)

		// NOTE: launch() sets the observer pointer at the very end of starting
		// up all components successfully. While not ideal, this is the only
		// signal we have to report whether Hubble was initialized successfully
		// for now.
		if h.observer.Load() == nil {
			return errors.New("Hubble launch failed")
		}

		return nil
	}))

```

Reply:
@mhofstetter ahh thank you I wasn't sure, applied your suggestion [here](https://github.com/cilium/cilium/compare/180a89d527750717ddfe8140e0ca2f6d0b2bbf1b..3cf6e3affbd196ebaa33bcc32689db115f3fc1d7).

Reply:
btw: an alternative way of checking would be to change the signature of `launch` and return an error/bool and check for that instead of checking internals (e.g. `observer`)

Reply:
Yes I thought of it and started this way but I'd really like to avoid modifying anything in `launch()` for now.